flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/utils/AsyncUtil.java

```diff
@@ -0,0 +1,216 @@
+import java.util.function.Predicate;
+
+/** Contains utilities for AsyncScalarFunctions. */
+public class AsyncUtil {
+
+    private static final EmptyResponseResultStrategy EMPTY_RESPONSE =
+            new EmptyResponseResultStrategy();
+    private static final Predicate<Throwable> ANY_EXCEPTION = new AnyExceptionStrategy();
+
+    /**
+     * Checks whether it contains the specified kind of async function call in the specified node.
+     *
+     * @param node the RexNode to check
+     * @return true if it contains an async function call in the specified node.
+     */
+    public static boolean containsAsyncCall(RexNode node) {
+        return node.accept(new FunctionFinder(true, true));
+    }
+
+    /**
+     * Checks whether it contains non-async function call in the specified node.
+     *
+     * @param node the RexNode to check
+     * @return true if it contains a non-async function call in the specified node.
+     */
+    public static boolean containsNonAsyncCall(RexNode node) {
+        return node.accept(new FunctionFinder(false, true));
+    }
+
+    /**
+     * Checks whether the specified node is the specified kind of async function call.
+     *
+     * @param node the RexNode to check
+     * @return true if the specified node is an async function call.
+     */
+    public static boolean isAsyncCall(RexNode node) {
+        return node.accept(new FunctionFinder(true, false));
+    }
+
+    /**
+     * Checks whether the specified node is a non-async function call.
+     *
+     * @param node the RexNode to check
+     * @return true if the specified node is a non-async function call.
+     */
+    public static boolean isNonAsyncCall(RexNode node) {
+        return node.accept(new FunctionFinder(false, false));
+    }
+
+    /**
+     * Gets the options required to run the operator.
+     *
+     * @param config The config from which to fetch the options
+     * @return Extracted options
+     */
+    public static AsyncUtil.Options getAsyncOptions(ExecNodeConfig config) {
+        return new AsyncUtil.Options(
+                config.get(ExecutionConfigOptions.TABLE_EXEC_ASYNC_SCALAR_BUFFER_CAPACITY),
+                config.get(ExecutionConfigOptions.TABLE_EXEC_ASYNC_SCALAR_TIMEOUT).toMillis(),
+                AsyncDataStream.OutputMode.ORDERED,
+                getResultRetryStrategy(
+                        config.get(ExecutionConfigOptions.TABLE_EXEC_ASYNC_SCALAR_RETRY_STRATEGY),
+                        config.get(ExecutionConfigOptions.TABLE_EXEC_ASYNC_SCALAR_RETRY_DELAY),
+                        config.get(ExecutionConfigOptions.TABLE_EXEC_ASYNC_SCALAR_MAX_ATTEMPTS)));
+    }
+
+    /** Options for configuring async behavior. */
+    public static class Options {
+
+        public final int asyncBufferCapacity;
+        public final long asyncTimeout;
+        public final AsyncDataStream.OutputMode asyncOutputMode;
+        public final AsyncRetryStrategy<RowData> asyncRetryStrategy;
+
+        @JsonCreator
```

Comment:
Do we need a dedicated JSON serialization for this? Can't we just restore the instances using the properties from `ExecNodeConfig`? I think we do this already, right?